3x faster digits10 via binary search (speeds up $ for numbers)

timotheecour opened the pull request:
* dragonbox.decimalLength and schubfach.decimalLength are now refactored to reuse `digits10`
* `digits10` (used by `addInt`) is now 3x-4x faster
* `addInt(int|int64)` is now a bit faster
* allows a more efficient addInt(uint64) on top of this PR, see benchmarks https://github.com/nim-lang/Nim/pull/18324#issuecomment-868106938

## future work
- [ ] add `addInt` overload for uint|uint64

timotheecour committed:
3x faster digits10

ringabout commented on tests/benchmarks/tdigits10.nim:
```diff
@@ -0,0 +1,88 @@
+benchmark for digits10
+]#
+
+func digits10v1(num: uint64): int {.noinline.} =
```
Did you try removing noinline pragma?

timotheecour replied:
yes, see last commit + updated benchmark

ringabout replied:
Thanks

ringabout commented:
Maybe you can also optimize decimalLength or so procs in dragonbox and subfetch modules?

timotheecour committed:
add test

timotheecour commented:
> Maybe you can also optimize decimalLength or so procs in dragonbox and subfetch modules?

i added those TODO for future work; how about merging this PR first and then your PR https://github.com/nim-lang/Nim/pull/18318 can do those refactorings since that was the direction it was taking?

ringabout commented:
> i added those TODO for future work; how about merging this PR first and then your PR #18318 can do those refactorings since that was the direction it was taking?

I agree.

timotheecour committed:
fixup

ringabout commented on lib/system/strmantle.nim:
```diff
@@ -25,34 +25,38 @@ const digitsTable = "0001020304050607080910111213141516171819" &
-  else:
-    result = 12 + digits10(num div 1_000_000_000_000'u64)
+func digits10(x: uint64): int {.inline.} =
+  if x >= 10000000000'u64: # 1..10, 11..20
```
Nitpicking: Can you use thousand separators among the number to make it more clear?

timotheecour replied:
done

ringabout commented on lib/system/strmantle.nim:
```diff
@@ -25,34 +25,38 @@ const digitsTable = "0001020304050607080910111213141516171819" &
+    if x >= 10000000'u64: # 6..7, 8..10
+      if x >= 100000000'u64: # 8, 9..10
+        if x >=  1000000000'u64: 10 # 9, 10
+        else: 9
```
Spaces?

ringabout commented on lib/system/strmantle.nim:
```diff
@@ -25,34 +25,38 @@ const digitsTable = "0001020304050607080910111213141516171819" &
+    else: 6
+  elif x >= 100'u64: # 1..2, 3..5
+    if x >= 1000'u64: # 3, 4..5
+      if x >=  10000'u64: 5 # 4, 5
```
Ditto?

timotheecour committed:
address comment

timotheecour committed:
fixup

Araq commented:
For a general purpose library optimizing for small numbers is wiser than optimizing for one billion. But of course these aspects can be combined, use binary search after you know the number is bigger than 10_000, for example.

c-blake commented:
`digits10` is not exported.  Presumably the interest in optimizing it is `addInt`.  You do not need a precise bound here and copying a few words is one of the fastest things a CPU can do.

More explicitly, instead of making `digits10` faster with (as Araq wisely "almost says") some log-10-scale series of perhaps quite unpredictable branches, it is almost surely better to _**eliminate digits10**_  entirely.  The max length you need is easy to bound (20 bytes or so, I think).  Stack allocate that with some `array[24, char]` or whatever, generalize `numToString` to work with an `openArray[char]`, run it on the temp buffer and then extend the target & copy.

While I have not timed this, intuitively it should be faster in more general CPU settings.  It is certainly simpler from a coding perspective.

timotheecour commented:
> For a general purpose library optimizing for small numbers is wiser than optimizing for one billion. But of course these aspects can be combined, use binary search after you know the number is bigger than 10_000, for example.

@Araq  this is just a matter of re-arranging the intervals in the binary search; the binary search is still the fastest way strategy, no need to fall back to another strategy.

here's a variant that splits unequally the 1st interval:
1..5, 6..20
and then all subsequent intervals are equally split

the results show not much difference (within plus or minus 20% depending on the ranges of numbers considered, still much faster than the linear scan before this PR), even if constraining for small values (because of CPU branch prediction working as intended)

so I'd suggest going for the simplest approach as done in this PR, which doesn't impose a bias and doesn't try to outsmart the CPU branch predictor.

```nim
proc firstPow10(n: int): uint64 {.compileTime.} =
  case n
  of 1: 1'u64
  of 2: 10'u64
  of 3: 100'u64
  of 4: 1000'u64
  of 5: 10000'u64
  of 6: 100000'u64
  of 7: 1000000'u64
  of 8: 10000000'u64
  of 9: 100000000'u64
  of 10: 1000000000'u64
  of 11: 10000000000'u64
  of 12: 100000000000'u64
  of 13: 1000000000000'u64
  of 14: 10000000000000'u64
  of 15: 100000000000000'u64
  of 16: 1000000000000000'u64
  of 17: 10000000000000000'u64
  of 18: 100000000000000000'u64
  of 19: 1000000000000000000'u64
  of 20: 10000000000000000000'u64
  else:
    doAssert false
    0'u64

func digits10v3(x: uint64): int {.inline.} =
  if x >= firstPow10(6): # 1..5, 6..20
    if x >= firstPow10(13): # 6..12, 13..20
      if x >= firstPow10(17): # 13..16, 17..20
        if x >= firstPow10(19): # 17..18, 19..20
          if x >= firstPow10(20): 20 # 19, 20
          else: 19
        elif x >= firstPow10(18): 18 # 17, 18
        else: 17
      elif x >= firstPow10(15): # 13..14, 15..16
        if x >= firstPow10(16): 16
        else: 15
      elif x >= firstPow10(14): 14
      else: 13
    elif x >= firstPow10(9): # 6..8, 9..12
      if x >= firstPow10(11): # 9..10, 11..12
        if x >= firstPow10(12): 12 # 11, 12
        else: 11
      elif x >= firstPow10(10): 10
      else: 9
    elif x >= firstPow10(7): # 6, 7..8
      if x >= firstPow10(8): 8
      else: 7
    else: 6
  elif x >= firstPow10(3): # 1..2, 3..5
    if x >= firstPow10(4): # 3, 4..5
      if x >= firstPow10(5): 5
      else: 4
    else: 3
  elif x >= firstPow10(2): 2
  else: 1
```

> While I have not timed this, intuitively it should be faster in more general CPU settings. It is certainly simpler from a coding perspective.

feel free to write a PR implementing your idea and show a benchmark, until then, this PR speeds digits10 by ~3X.

> Presumably the interest in optimizing it is addInt

dragonbox and schubfach also use the same API, modulo refactoring, so it also helps for addFloat

>  it is almost surely better to eliminate digits10 entirely. 

how would $x work so that it returns a minimum length string without digits10?

c-blake commented:
Your dismissal of my argument as trying to not outsmart the branch predictor or having a bias is invalid.  What bias?  Giving something much less work to do is not "outsmarting it"; it's helping it.  BPs are not magic and binary search is a known problem area for them (search for it if you do not trust me).  Here it is search for digit length where if you only have 3-4 numbers of digits instead of the full 19 range you may not see effects.  The max depth is probably 5 an almost never more than 4.  Beats me what your benchmark is.  Anyway, zero time is infinitely faster than whatever it costs and your search code is horrendously complex, as I mention to detriment not benefit over a simple alternative.

Sorry, but this direction is enough off track for a simple int to string conversion that I think closing this PR to force you to actually try a simpler, almost surely better idea might even be warranted.  Up to @Araq of course, but that's my advice.  Cheers.

c-blake commented:
Don't get me wrong - if you have some *other* reason for a `ceil(log_10(x))` operation then it might be appropriate, but int -> string where you only need to bound buffer size is simply not a good reason.  All the linked to resources I found on this from this PR point to other people who are similarly confused.

timotheecour commented:
> it is almost surely better to eliminate digits10 entirely
> [....] but int -> string where you only need to bound buffer size is simply not a good reason

uhm, how would you set the string length without `digits10`? Or correctly compute the minimum string representation in addFloat (dragonbox relies on it too, modulo refactor) ? bounding a buffer size isn't what this is about.

> What bias? 

biasing towards small sized inputs ("optimizing for small numbers is wiser than optimizing for one billion"). Most FP numbers fed through dragonbox.decimalLength yield a number of digits that is in the full range.

```nim
# dragonbox.nim:
proc decimalLength*(v: uint64): int32 {.inline.} =
  defer: debugEcho result, " ", v # add this for debugging
  ...
```

```nim
when true:
  proc fn(a: float) =
    let s = $a
  fn(1.2)
  fn(10.1)
  fn(10.0)
  fn(1e8)
  fn(10/3)
```

output:

16 1200000000000000
16 1010000000000000
2 10
9 100000000
17 33333333333333335

=> no small numbers

c-blake commented:
To reply to the rather impolite *late edited* in:
> Feel free to add a PR
> bounding a buffer size isn't what this is about.

On the contrary, in an integer setting that bound is exactly what this is about.  The only trick with int -> str is that it creates the answer from the back of the string forward.  See `numToString`.  So, you alloc a big enough temp buffer which is easy in this case from a representation bound on uint64.  The number of digits is an implicit result of the calculation you must do anyway - whenever you write the last byte into the produced buffer going backwards, you know the address of that byte and you had to know the address of the end to get started.  So, the diff between the end of string and that byte (+ 1) is your length.  In the temp buffer setting you can then just ensure room in the target & copy.  That's it!  You just do not need to know the number of digits in advance for int -> str if you are willing to allocate a 20 byte buffer on the stack, and I don't see why you wouldn't be willing.  That's like 3 int local vars.

This might make for a slightly different internal numToString interface.  That did not seem to be a point of contention.  The point here is to "back away" slightly from the problem instead of optimizing an unneeded abstraction of pre-knowledge of the number of digits.

I can't think of a way to explain this any more simply.  I am already mostly repeating myself here, but Araq's heart makes it seem worthwhile.  (EDIT: The issue of bias to small numbers simply disappears in my idea.  The only bias is using less of an eentsy line stack buffer which is to say, no performance bias other than the ineliminable number of loops to do the work "bias".)

c-blake commented:
Note - I have not evaluated (or claimed!) whether this advance knowledge is really needed for dragonbox, although I would be skeptical up front.  If dragonbox does need it, then it can use this big mess over there.  That doesn't mean *this* integer context cannot be simply this with `digits10` deleted (and a version for negatives added, templates, etc.).  Then both parts will be fast and only as complex as they need to be (floats are complex - ints are not..why burden the latter with the former - optimization is specialization.  So, specialize where it's actually desirable in the first place):
```Nim
proc add(target: var string, x: uint64) =
  const digit = "0123456789"          # maybe set elsewhere already
  var buf {.noInit.}: array[20, char] # ceil(log_10(2^64 - 1)) == 20
  var x = x                           # changing perms on `x`
  var i = buf.len - 1                 # work from back of buf to front
  while x >= 10:
    buf[i] = digit[x mod 10]  # extract digit into temp array
    x = x div 10
    dec i
  buf[i] = digit[x mod 10]
  x = x div 10                # No final dec; buf[i] is now leading digit
  let n = target.len          # a couple trivial abstractions
  let m = buf.len - i         # could also just target.add buf[i..^1]
  target.setLen n + m         # this all works in base-100, too
  copyMem target[n].addr, buf[i].addr, m

when isMainModule:
  var s = ""
  s.add 1  ; echo s.repr
  s.add 22 ; echo s.repr
  s.add 333; echo s.repr
```
EDIT: and sure, you can squeeze some more perf out by doing digits in batches of 2 (base 100, essentially to do fewer divmod's), but you *still* do not need/want the whole number of digits in advance.  So, it is dubious this work advances future integer work either.  Seems to me like making a mess where a clean-up could happen instead.  Saying anything more would just be repeating myself which I've done too much already.

Araq commented:
@c-blake Please create an alternative PR.

c-blake commented:
It'll have to wait until at least this coming weekend.  If anyone else wants to, _**please**_ feel free.

timotheecour committed:
debugging

timotheecour committed:
fixup after rebase

timotheecour committed:
faster

timotheecour committed:
PRTEMP

timotheecour committed:
cleanups

timotheecour committed:
cleanup benchmark

timotheecour committed:
cleanups

timotheecour committed:
avoid using toOpenarray because of #15952

timotheecour commented:
PTAL
* dragonbox.decimalLength and schubfach.decimalLength are now refactored to reuse digits10
* regarding removing `digits10` as advocated by @c-blake:
`digits10` was introduced in https://github.com/nim-lang/Nim/pull/16160 to speedup addInt and @c-blake's approach is similar to the one we had prior to that PR. Furthermore, `digits10` is also needed for dragonbox and schubfach.

I've created a benchmark in https://github.com/timotheecour/vitanim/blob/master/testcases/t10977.nim comparing @c-blake's approach from https://github.com/nim-lang/Nim/pull/18324#issuecomment-865947585 (named `addInt1`) to addInt based on `digits10` (named `addInt2`), from this PR, defined as:
```nim
proc addInt2(result: var string; x: uint64) =
  let length = digits10(x)
  setLen(result, result.len + length)
  addIntImpl(result, x, result.len - length)
```

results show that `addInt2` is faster, here by 1.6X:
(2.589571, "addInt1", 974118343)
(1.624844, "addInt2", 974118343)

the benchmarks also has several distributions of pseudo-random uint64's defined (eg a distribution favoring smaller ints, different pseudo-random variants); `addInt2` is consistently faster than `addInt1`.

I also tried with to speedup @c-blake's version using the same optimization we use (1 call to mod 100 instead of 2 calls to mod 10), it does improve the speed, but `addInt2` is still faster.

In future work, I will add an `addInt` overload for uint|uint64, defined as above (the fact we don't have it is just an omission in stdlib).

c-blake commented:
To be clear, my code above was just to respond to weird points about how it was even possible which seemed implausibly in dispute.  Raw code seemed the only remaining way to make it more clear. At least we're passed that now.  So, maybe it was necessary.

Of course half the loops is faster as mentioned in my edit there..Curious you don't tell us that number.  One might think applying "BS scaling" that about half the loops taking that 1.6x down to 0.8x or faster than your idea, but not *all* work is the divmods/loop count.  So, your "not as fast" is credible, but I would guess its only 5..25%.  If so I would say the extra code & CPU complexity/I-cache/uop cache is likely unwarranted by your very own measurement.

A delta that small is also virtually guaranteed to not be CPU/compiler/compiler-flag/PGO/--gc:/etc. portable.  It may well be over-tuned to whatever environment you are testing on.  I usually think +-25% is basically such noise, but it can surely spike higher.  I also feel like I've heard @Araq mention a figure of near 25% as being "silly" more than once (possibly from over-tuned to one deployment experiences in the past - hey, we've all been there!). Point is +25% in one deployment can easily be -25% somewhere else.

I don't know what to say about the 2.6x xflywind found, but it didn't look like he even checked more than one CPU microarch/backend compiler/GC framework.  So, you know, partially informative at best.  In general I think all these benchmarks not reporting at least a few CPUs/couple compilers is actually kind of an irresponsible way to measure for a stdlib with diverse deployments.  Code complexity without responsible justification is another kind of irresponsibility.  I think we can/should do better.

But I will do my own PR this weekend or so, as I mentioned.  I just had to say your own benchmark is probably not telling a great tale here for advance digits knowledge.  Even 1.6x is barely motivating.  It seems it will be down to much finer +- 25% deployment sensitive hair splitting.  For speed hair splitting, simplest in both code & processor resources wins, IMO.  I wouldn't think that either contentious or surprising.

Believe it or not, I do not try to make contentious claims.  I would honestly find it rare "in the wild" for anyone knowing CPU basics to believe any binary search is less work than a 3 register stack alloc & copy.  The extent to which any of that winds up a net wash here may well be from (*some*) CPUs (in *some* circumstances) being able to interleave early dispatched alloc work with near future calculation -- a fragile state of affairs, indeed.  It may be seeming to do "more work faster", but then as soon as other things start to happen in real user code the "more work come uppance" will rear its ugly head.  Complex code yielding fragile perf - not good.

Here's another non-contentious one - a base1000 3k buffer or base10000 40k buffer might be even faster for a hot loop microbenchmark but at even more significant competing costs making constructing representative benchmarks even harder.  At 40k you are even into binary executable size tradeoffs (or run-time startup population ones).  But, yeah, in a hot loop the mostly L1 loads will still beat divmod.  Heck, for low enough entropy number digits/sizes you could probably even get a 6*1e6 L3 table do better with almost always <3 loops.

So, you know, I kind of think how to do `numToString` for ints is mostly about hard to assess deployment-stable performance considerations as well as code complexity for machine & man.  But over in dragonbox where probably because users tell the routine the number of digits it helps to have advance knowledge, it may be reasonable.  That module is also already crazy complex, but also in the service of more than just (weakly measured) perf.  Cheers.

ringabout commented:
> I don't know what to say about the 2.6x xflywind found, but it didn't look like he even checked more than one CPU microarch/backend compiler/GC framework. So, you know, partially informative at best. In general I think all these benchmarks not reporting at least a few CPUs/couple compilers is actually kind of an irresponsible way to measure for a stdlib with diverse deployments. Code complexity without responsible justification is another kind of irresponsibility. I think we can/should do better.

Sorry, I was irresponsible at that time because I didn't have much experience and nim language/programming language was still a new thing to me. Having seen that other languages tried this trick, so I decided to port this to Nim.

Back then, I sought some advice from @timotheecour multiple times in our private gitter chat. Then at least Windows and MacOS are checked.

Clyybber commented:
Running the benchmark that @timotheecour posted (made runnable without checking out this PR): https://play.nim-lang.org/#ix=3qSu on my machine gives
```
(1.6764043499999999, "addInt1", 974118343)
(1.5283274910000006, "addInt2", 974118343)
```
for the first case `# pseudo-random in 0..M`
```
(2.363221254, "addInt1", 1718245140)
(1.894901105999999, "addInt2", 1718245140)
```
for the second case `# pseudo-random variant 2`, and
```
(1.340069962, "addInt1", 764000000)
(1.311113476, "addInt2", 764000000)
```
for the third case `# pseudo-random with small number bias`.
The most extreme difference of these results is a ~1.25 difference, coincidentally exactly the +-25% @c-blake describes :)

c-blake commented:
@xflywind - to be clear, I'm just trying to encourage better measuring/motivation so that we don't a year or whatever later have to say "how could that 2.6x" have happened (EDIT: and for all I know it is reproducible now...).  I realize it may have come off as overly harsh.  FWIW, that Andrei article you guys liked missed the base100 elephant in the room optimization.

@Clyybber - thanks for following up/confirming.  It looks like you are doing the 1-digit at a time not 2-digit at a time no-advance digits10.  The 2-digit should be even closer or even better (and is more apples to apples).

I really do think 3 reg push & 3 reg copy is fundamentally "less work" than a big flattened binary search. "Direct timing in specific benchmarks" might be pretty confusing here as much as everyone tilts toward `for i in 0..<1000000:...`.  Less memory/less work/more predictable work == good optimization.  Big complex branchy things that might eek out a few % in some specific situation == unlikely to be good/survive/portable/enduring.  Anyway, more in a few days.  Cheers all.

Araq commented:
Well if it doesn't harm other test cases (and of course we can never say that for certain) a 25% performance improvement is nothing to complain about.

c-blake commented:
If it's a *consistent* gain across deployments -- which is in doubt.  Preliminary tests I did showed 3/4 CPUs *slower* with the big branchy thing (by 7..25%).

timotheecour commented:
here's the numbers I'm getting with the snippet https://play.nim-lang.org/#ix=3qSu posted by @Clyybber in https://github.com/nim-lang/Nim/pull/18324#issuecomment-866895695

params for reproducibility:
* OSX 11.4, 2.3 GHz 8-Core Intel Core i9
* using `nim r -d:danger --skipusercfg --skipparentcfg main` for all 3 benchmarks
* using nim 9a81e91fa5d01cd6d60a408d0888dce4a9f49898 (but shouldn't matter)

## benchmark for addInt
pseudo-random in 0..M
(2.654544, "addInt1", 974118343)
(1.6393009999999997, "addInt2", 974118343)
=> speedup 1.6X

pseudo-random variant 2
(3.459155, "addInt1", 1718245140)
(1.8157, "addInt2", 1718245140)
=> speedup 1.9X

pseudo-random with small number bias
(2.2029539999999996, "addInt1", 764000000)
(1.490523, "addInt2", 764000000)
=> speedup 1.48X

@Clyybber did you run with -d:danger? and which machine did you run on? I'm curious whether the large difference observed in your timing results is due to the difference in CPU or whether it's something else.

Can other people run this benchmark too (with `nim r -d:danger --skipusercfg --skipparentcfg main`) ? it's just a matter of copy pasting and running


## benchmark for just digits10
what results do you get for `nim r -d:danger tests/benchmarks/tdigits10.nim` ? is it consistent with the results i'm getting (as noted in PR):
```
("digits10", 18976262687, 1.685551) # new
("digits10v2", 18976262687, 6.813552) # old
```
=> speedup 4x

## note
--skipusercfg --skipparentcfg to avoid your local environment affecting outcome, but shouldn't normally matter

c-blake commented:
As already mentioned, the @Clyybber snippet does not do 2-digits at a time.  So some baseline BS scaling of 2X faster is in play which is bigger than the above ratios (but yeah, BS).  So, I don't think a request for other people to run that is very worthwhile because I do not think this should be the real idea comparison as to whether digits10 is good.

Your own mentioned attempt at a 2-digit-at-a-time no advance digits

>I also tried with to speedup @c-blake's version using the same optimization we use (1 call to mod 100 instead of 2 calls to mod 10), it does improve the speed, but addInt2 is still faster.

would be far more worthwhile.  Or literally even putting numbers on that last "still 'faster'" or posting that code.  I already called this out, and already said - if people are *patient* - that I will do something.  Or @Clyybber could try the real 2 digit thing before worrying about checking his environment to triple check the strawman comparison (which will never decide this particular `digits10` for ints question).

Clyybber commented:
> @Clyybber did you run with -d:danger? and which machine did you run on? I'm curious whether the large difference observed in your timing results is due to the difference in CPU or whether it's something else.

`--gc:arc -d:danger` running on a Ryzen 7 4800H. 

>  Or @Clyybber could try the real 2 digit thing before worrying about checking his environment to triple check the strawman comparison (which will never decide this particular `digits10` for ints question).

Agreed, it doesn't seem to be a fair comparison.

Clyybber commented:
I added @c-blake's 2 digit version (addInt3) to the snippet https://play.nim-lang.org/#ix=3qUT and on my machine it is consistently slightly faster than addInt2.

timotheecour commented:
results I get with https://github.com/nim-lang/Nim/pull/18324#issuecomment-867123234:

nim r -d:danger -d:case1 test2
(1.683055, "addInt2", 974118343)
(1.759594, "addInt3", 974118343)
=> addInt2 1.045X faster

nim r -d:danger -d:case2 test2
(1.879833, "addInt2", 1718245140)
(2.069533, "addInt3", 1718245140)
=> addInt2 1.1X faster

nim r -d:danger -d:case3 test2
(1.505091, "addInt2", 764000000)
(1.6452350000000002, "addInt3", 764000000)
=> addInt2 1.1X faster

nim r -d:danger --gc:arc -d:case1 test2
(1.720806, "addInt2", 974118343)
(1.640957, "addInt3", 974118343)
=> addInt2 1.048X slower

nim r -d:danger --gc:arc -d:case2 test2
(1.8232540000000002, "addInt2", 1718245140)
(1.7848500000000003, "addInt3", 1718245140)
=> addInt2 is 1.02X slower

nim r -d:danger --gc:arc -d:case3 test2
(1.417223, "addInt2", 764000000)
(1.388094, "addInt3", 764000000)
=> addInt2 is 1.02X slower

In my version of addInt3 (mod 100 without digits10), I had only tested -d:danger, not -d:danger --gc:arc, and I had observed that addInt2 was still a little bit faster than addInt3. But with `-d:danger --gc:arc` (which I hadn't tried in my earlier benchmark), it's the opposite, addInt3 is a little bit faster than addInt2 (by a smaller factor but let's not split hair). The reason I went for digits10 instead of addInt based on `mod 100 without digits10` is that AFAIK we still need digits10 anyways for dragonbox+schubfach, so that we don't have too many distinct code paths by reusing code. 

Given the small performance differences between addInt2 and addInt3 (both with or without --gc:arc), I'm ok with the following:
* implement addInt without digits10 (using mod 100)
* but keep the rest of this PR, including the modification to digits10, which is used in dragonbox+schubfach, and can be useful on its own

Araq commented:
You all lost me but I have faith in our process and that eventually the best solution wins. :-)

c-blake commented:
A) `addInt` performance evaluation was pretty poor and is basically withdrawn as justification.

B) I see no evaluation of perf for `digits10` at all *in call context* for the two other 2 potential use cases which each uses its own `digits10` specialization anyway (32-bit for schubfach & 64-bit for dragonbox).  So certainly no I-cache saving.

As a prior for why that evaluation should precede PR acceptance of anything in this space, at least dragonbox came from a source seemingly concerned with speed.  I see no reason to think you are a better benchmarker than them.  Their code has ***far*** wider exposure than just a Nim PR.  Maybe they missed this optimization - ***or maybe*** they tried on a broad set of CPUs or broad real user input and found their scan worked better.

Beyond this, both algos seem to use this digit count procedure on decimal significands of binary floats. Those strike me as likely to be statistically strongly biased towards max length ints/very few loops counting down.  So "bias" here is probably towards super big, not small.  Unless you have special low entropy float sources (perfect binary fractions like .125), linear scan from the largest might only do a branch or three and be binary search.  Like so much, this one probably just depends on inputs, but I doubt an optimization for special low entropy float inputs will be a very popular/common need.

Even without both priors & questions, CPUs are complex enough that it is simply very wrong to just assume it will help in context based on a microbenchmark and press to merge this based on un-done follow on work that may well fail.

C) DRY for its own sake is also no rationale, especially in an optimization setting.  Given B), there may be literally ***no other use case*** for a one-size-fits all pair of `digits10` s -- if one *even exists*.  It seems (roughly) agreed upon that it is an easily fixed mistake in `numToString`.  Why export it at all?  Leave them defined locally in the two relevant modules where specialization & optimization can peacefully coexist.  Sure seems premature to think you have found (or even ***can***) the best ever `digits10*` for both use cases (***or*** all 3 cases before `addInt` debate, really -- ***or*** some unanswered 4th possible case).

In short, `digits10*` seems like misargued, impulsive factoring that may well see no client use in the stdlib, just being a big unused mess, exported even and ugly to boot!  Nim needs no more of this.  Even with no competing PR, this one (and its hoped for follow on "fixes") are bad for later stage optimization in an area where people complain about speed.  @Araq - I don't know what's confusing about literally any of this if you think about even just (C) for like 5 minutes.

Araq commented:
> I see no reason to think you are a better benchmarker than them. Their code has far wider exposure than just a Nim PR. Maybe they missed this optimization - or maybe they tried on a broad set of CPUs or broad real user input and found their scan worked better.

Too much of an "argument by authority" for me. If Timothee can show his ideas do improve the performance then that's what he does, regardless of who the original authors of the code were.

c-blake commented:
That part was only about priors of actually showing improvement before accepting PRs.  Apologies if it was unclear.  Timothee seems to want this PR based on future, unshown improvements.

timotheecour commented:
@Araq  After investigation, the performance differences observed with gc:arc|refc are due to https://github.com/nim-lang/Nim/issues/18339 and gc performance issues for strings and added noise and confusion to the measurements.

Here's an updated and improved benchmark where I take the gc out of the equation:
https://github.com/timotheecour/vitanim/blob/master/testcases/t10977b.nim
results as of 5a317b0879f73e9aa9313ce910cb38301891cded, using nim 16038d44f6b246c5f39d1a8609a8c45f2cd005a4, on OSX 11.4, 2.3 GHz 8-Core Intel Core i9:

I'm comparing 3 distributions (see test for details):
* mode 1 = pseudo-random in 0..M (via a recursion)
* mode 2 = pseudo-random in 0..M (via a recursion + iteration index)
* mode 3 = pseudo-random with small number bias, so that digits of all lengths are likely

I'm comparing 5 algorithms:
* addInt0: naive algo = best you can do in current nim (which lacks addInt for uint64), via `$x`
* addInt1: proposed in https://github.com/nim-lang/Nim/pull/18324#issuecomment-865947585 by @c-blake (mod 10 without digits10)
* addInt3: proposed in https://github.com/nim-lang/Nim/pull/18324#issuecomment-867123234 by @Clyybber (mod 100 without digits10)
* addInt3OpenArray: I took addInt3 and modified it to work with openArray to avoid performance issues with string's
* addIntDigits10OpenArray: this PR's addInt modified to work with openArray and improved a bit:
```nim
proc addIntDigits10OpenArray(ret: var openArray[char]; num: uint64): int {.inline.} =
  result = digits10(num)
  var x = num
  var i = result - 2
  while i >= 0:
    let xi = (x mod 100) shl 1
    x = x div 100
    copyMem ret[i].addr, digits100[xi].unsafeAddr, 2
    i = i - 2
  if i == - 1: ret[0] = chr(ord('0') + x)
```

here are the results, showing `addIntDigits10OpenArray` is consistently faster than all other algorithms, both with gc:arc|refc; in particular it's:
* up to 1.76X faster than addInt3OpenArray
* up to 3.6X faster than addInt3
* up to 4.4X faster than addInt1
* up to 8X faster than addInt0

```
# nim r -d:danger $vitanim_D/testcases/t10977b.nim
addInt0
(time: 3.960433, alg: "addInt0", mode2: 1, sanity: 6057706836)
(time: 4.495959999999999, alg: "addInt0", mode2: 2, sanity: 6823134569)
(time: 4.215945000000001, alg: "addInt0", mode2: 3, sanity: 5913000000)

addInt1
(time: 2.472244999999999, alg: "addInt1", mode2: 1, sanity: 6057706836)
(time: 3.4021859999999986, alg: "addInt1", mode2: 2, sanity: 6823134569)
(time: 2.3408859999999976, alg: "addInt1", mode2: 3, sanity: 5913000000)

addInt3
(time: 1.8768990000000016, alg: "addInt3", mode2: 1, sanity: 6057706836)
(time: 2.1484859999999983, alg: "addInt3", mode2: 2, sanity: 6823134569)
(time: 1.9108490000000025, alg: "addInt3", mode2: 3, sanity: 5913000000)

addInt3OpenArray
(time: 1.007784000000001, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 1.329969000000002, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 0.9360080000000011, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)

addIntDigits10OpenArray
(time: 0.7906410000000008, alg: "addIntDigits10OpenArray", mode2: 1, sanity: 6057706836)
(time: 0.8743730000000021, alg: "addIntDigits10OpenArray", mode2: 2, sanity: 6823134569)
(time: 0.5313129999999973, alg: "addIntDigits10OpenArray", mode2: 3, sanity: 5913000000)
```

ringabout commented:
I ran @timotheecour's benchmark on my win10 machine, here is the result:

nim r -d:danger test.nim
```
addInt0
(time: 10.364, alg: "addInt0", mode2: 1, sanity: 6057706836)
(time: 10.116999999999999, alg: "addInt0", mode2: 2, sanity: 6823134569)
(time: 8.439, alg: "addInt0", mode2: 3, sanity: 5913000000)

addInt1
(time: 3.187000000000001, alg: "addInt1", mode2: 1, sanity: 6057706836)
(time: 4.854999999999997, alg: "addInt1", mode2: 2, sanity: 6823134569)
(time: 2.8260000000000005, alg: "addInt1", mode2: 3, sanity: 5913000000)

addInt3
(time: 2.644999999999996, alg: "addInt3", mode2: 1, sanity: 6057706836)
(time: 3.4080000000000013, alg: "addInt3", mode2: 2, sanity: 6823134569)
(time: 2.4080000000000013, alg: "addInt3", mode2: 3, sanity: 5913000000)

addInt3OpenArray
(time: 1.5940000000000012, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 2.2439999999999998, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 1.4779999999999944, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)

addIntDigits10OpenArray
(time: 1.0870000000000033, alg: "addIntDigits10OpenArray", mode2: 1, sanity: 6057706836)
(time: 1.8220000000000027, alg: "addIntDigits10OpenArray", mode2: 2, sanity: 6823134569)
(time: 1.1029999999999944, alg: "addIntDigits10OpenArray", mode2: 3, sanity: 5913000000)
```

Nim version:
```
Nim Compiler Version 1.5.1 [Windows: amd64]
Compiled at 2021-06-19
```
System info:
```
4core + 8G
Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz   2.90 GHz
```

ringabout commented:
Run it on my ubuntu system(cloud server)

```
addInt0
(time: 7.580807493, alg: "addInt0", mode2: 1, sanity: 6057706836)
(time: 9.485953271, alg: "addInt0", mode2: 2, sanity: 6823134569)
(time: 7.489823507000001, alg: "addInt0", mode2: 3, sanity: 5913000000)

addInt1
(time: 3.699648988, alg: "addInt1", mode2: 1, sanity: 6057706836)
(time: 5.147743381999998, alg: "addInt1", mode2: 2, sanity: 6823134569)
(time: 3.348001267000001, alg: "addInt1", mode2: 3, sanity: 5913000000)

addInt3
(time: 3.075346617000001, alg: "addInt3", mode2: 1, sanity: 6057706836)
(time: 3.960488266999995, alg: "addInt3", mode2: 2, sanity: 6823134569)
(time: 2.826448000999996, alg: "addInt3", mode2: 3, sanity: 5913000000)

addInt3OpenArray
(time: 1.781527083, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 2.520438331000001, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 1.693915054000001, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)

addIntDigits10OpenArray
(time: 1.214152513000002, alg: "addIntDigits10OpenArray", mode2: 1, sanity: 6057706836)
(time: 2.042936627000003, alg: "addIntDigits10OpenArray", mode2: 2, sanity: 6823134569)
(time: 1.170955093000003, alg: "addIntDigits10OpenArray", mode2: 3, sanity: 5913000000)
```
system info： 1 core 2g
```
Linux iZuf60lvzs716yak68vciwZ 5.4.0-67-generic #75-Ubuntu SMP Fri Feb 19 18:03:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

model name      : Intel(R) Xeon(R) CPU E5-2682 v4 @ 2.50GHz
cpu MHz         : 2499.994
cache size      : 40960 KB
```

timotheecour committed:
improve

timotheecour committed:
improve addInt(int64)

Clyybber commented:
> Here's an updated and improved benchmark where I take the gc out of the equation:
https://github.com/timotheecour/vitanim/blob/master/testcases/t10977b.nim

Comment out
```nim
    mainAux(addInt0)
    mainAux(addInt1)
    mainAux(addInt3)
    mainAux(addIntDigits10)
```
and suddenly I get
```
addInt3OpenArray
(time: 0.900472838, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 1.2079957449999998, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 0.8015987450000002, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)

addIntDigits10OpenArray
(time: 0.6627606450000001, alg: "addIntDigits10OpenArray", mode2: 1, sanity: 6057706836)
(time: 4.987259771000001, alg: "addIntDigits10OpenArray", mode2: 2, sanity: 6823134569)
(time: 2.0311305530000006, alg: "addIntDigits10OpenArray", mode2: 3, sanity: 5913000000)
```

Just goes to show that we are really not measuring something useful here.

You can also look at the size of the generated assembler, if you still think that your version using digits10 should be intuitively or generally faster: https://godbolt.org/z/4Wjjz9Ebd

c-blake commented:
@xflywind - your CPU microarchitecture would be helpful.

I haven't looked into this new stuff in detail and these reports are getting harder to read except @Clyybber's whose was nicely abbreviated. What I'm hearing from TC & xfly is that allocation noise swamped your effect. Maybe that noise is other "mountains of work" *or maybe* it's actual CPU resource competition (which matters for *literally anything* besides your microbenchmark which maybe isn't even indicative anyway as per @Clyybber).  In either event, since `$` allocates the PR title misleads as this is leaning more about optimized `addInt`.

Many of us take microbench shortcuts as a first pass.  You can learn more by follow-on CPU profiling (in context) and watching implied time, looking at `perf stats`, etc. -- hopefully for at least a few programs on a few CPUs.  While always true, this can be an especially big "correction" with things like a 100 instruction-500 byte flattened binary search.  Resource competition exists.  uop caches are small.  I may seem too demanding here, but it isn't unreasoned.  As @Clyybber points out, that binary search is clearly much more code, more work, and work more contingent upon prediction.  Yet, the claim is "overall binsearch much faster!"  Now, this claim isn't as extraordinary as "The Earth is Flat" - the write pattern isn't identical and so doesn't need *as* extraordinary evidence.  I would sure like to hope that the disparity implies more measurement care/really solid evidence than I see without a lot of conflicting which is faster results with big ratios in either direction.

No matter what else, exporting `digits10` (private/ or not) with an intent to share is premature at best.  Please don't do that.  While seemingly born as a "lift abstraction out and optimize the crap out of it" scenario, you don't know that even works yet and there are good reasons to suspect it will not.  There are worst case vs. average case problems with a centralized `digits10` impl (the many digits worst case in FP is *super* common - eg divide by 3! but rare for integers outside `.addr`s).  Big O/worst case analysis became popular for a reason..

timotheecour commented:
> Comment out
> and suddenly I get
[...]

@Clyybber  after investigating, here's what i found:
## on osx:
* commenting those out makes no difference with gc:arc|refc
## on ubuntu (through docker):
* commenting those out makes no difference with gc:refc
* with gc:arc, something really strange happens:
whichever is the 2nd call gets slower by a large factor, 3.5X, even if the same algorithm is running:
```nim
proc main =
  for i in 0..<4:
    echo()
    mainAux(addInt3OpenArray, needResize = false) # A
    mainAux(addInt3OpenArray, needResize = false) # B
main()
```
regardless of what A and B are, B will be slower by a factor 3.5X in each loop iteration, even if both A and B are the same (eg addInt3OpenArray, which doesn't use digits10 at all)

in other words, in pseudo code:
```nim
for i in 0..<4:
  addIntDigits10OpenArray
  addInt3OpenArray # will be slower
```
```nim
for i in 0..<4:
  addInt3OpenArray 
  addIntDigits10OpenArray # will be slower
```
```nim
for i in 0..<4:
  addInt3OpenArray 
  addInt3OpenArray # will be slower
```

```
addInt3OpenArray
(time: 1.00154777, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 1.487595834, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 0.9652565010000003, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)

addInt3OpenArray
(time: 0.9771263480000001, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 5.496433305999999, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 3.292133801, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)


addInt3OpenArray
(time: 1.004118004, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 1.517611261000001, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 0.9857435569999993, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)

addInt3OpenArray
(time: 0.9622406769999969, alg: "addInt3OpenArray", mode2: 1, sanity: 6057706836)
(time: 5.562718226000001, alg: "addInt3OpenArray", mode2: 2, sanity: 6823134569)
(time: 3.322252200000001, alg: "addInt3OpenArray", mode2: 3, sanity: 5913000000)
...
```

so this looks like a strange performance issue with gc:arc at least on linux. Again, in this example, digits 10 is not used.

timotheecour commented:
@Araq  @Clyybber please try again this revision of the benchmark: 0a6069a73434533ef700e5a197419d56daf5c2bf from https://github.com/timotheecour/vitanim/blob/master/testcases/t10977b.nim
all it does is add a `block:`

I'm now observing the advertised speedups (with -d:danger) no matter:
* the os (osx, linux)
* the gc (arc, refc)
* whether or not you comment out the addInt* procs (https://github.com/nim-lang/Nim/pull/18324#issuecomment-868397137)
* whether or not you change ordering of the algos, including addInt3OpenArray vs addIntDigits10OpenArray

and it doesn't exhibit the non-sensical issues observed above in https://github.com/nim-lang/Nim/pull/18324#issuecomment-868684436 where the 2nd algo (even if same) was always 3.5X slower when running in gc:arc; that problem was unrelated to digits10 as it was reproduced without even digits10 being called, and is IMO a performance issue specific to arc that should be investigated independently.

ringabout commented:
I have checked the revised test on win10(-d:danger): 0a6069a73434533ef700e5a197419d56daf5c2bf from https://github.com/timotheecour/vitanim/blob/master/testcases/t10977b.nim

with refc:
addIntDigits10OpenArray has consistent improvement speedups compared to others.

using `-d:danger --gc:arc` seems not be affected by the bug mentioned @timotheecour, the result(arc) listed here:
https://gist.github.com/xflywind/fd6b7aa00767110d34e098186e8b3013

c-blake commented:
@xflywind - Not me.

On that latest test I get for the `addInt3OpenArray` vs `addIntDigits10OpenArray` that the simpler former one is faster for all but mode 3, and the digits-in-advance in mode3 is only 8% faster.  Also, I'm not sure who ever said "50" was a magic threshold for "small numbers".  100 is more natural in context with the 2-digit at a time algorithm.  If I change that to 9999 for 4 digits that 8% vanishes and digits-in-advance is slower in all 3 modes.  We need more than consistency on a couple CPUs.

Performance variation against [the proposed alternative](https://github.com/nim-lang/Nim/pull/18356) really is all about CPU microarch and even ratios as high as 2..3x are not so rare.  Why just yesterday I saw, for example, a variation from 60x to 20x over at https://github.com/nim-lang/Nim/issues/18339, and the "ratio of ratios" is 3X.  For this code, even minor rearrangments and compiler flag adjustments of this `$` stuff yield as big deltas as Timothee claims are "real".  At a bare minimum this means that users might have to "tune their flags" to get the fast `$` in a way *differently* from tuning them for the rest of their program.

Besides these considerations, if people in Nim land are going to be debating endlessly over 25..50% mystery CPU microarch effects then there should be some standardized set of test platforms, not whatever TC/xfly/cb are using this month, in the name of reproducibility & consistency if nothing else.

Varriount commented:
To add another viewpoint to the mix: if the performance difference is sufficiently small/unpredictable, I feel that the simpler of the two implementations should be used. Aside from readability, I would assume that a simpler implementation would hold up better across a larger number of architectures.

ringabout commented:
Since digit10 has used the other algorithm which is not compatible with this PR, closing for now.
